File: frontend/src/api/stats.ts
/**
 * Stats API functions.
 */

import { apiFetch } from "./client";

export interface DailyTaskStat {
  date: string;
  total: number;
  done: number;
}

export interface DailyMoodStat {
  date: string;
  mood: string;
}

export interface DailyWorkStat {
  date: string;
  minutes: number;
}

export interface WorkLabelStat {
  label: string;
  minutes: number;
}

export interface ProjectTimeStat {
  project_id: number | null;
  name: string;
  minutes: number;
}

export interface WeeklySummaryCounts {
  work_minutes: number;
  tasks_done: number;
  notes: number;
  feynman: number;
  moods: number;
}

export interface WeeklySummary {
  this_week: WeeklySummaryCounts;
  prev_week: WeeklySummaryCounts;
}

export interface UserStatsRead {
  days: number;
  daily_tasks: DailyTaskStat[];
  daily_moods: DailyMoodStat[];
  daily_work_minutes: DailyWorkStat[];
  work_labels: WorkLabelStat[];
  time_per_project?: ProjectTimeStat[];
  total_tasks_done: number;
  total_work_minutes: number;
  total_notes: number;
  total_feynman: number;
  total_moods?: number;
  weekly_summary?: WeeklySummary | null;
}

export interface UserProgressRead {
  user_id: number;
  xp: number;
  level: number;
  xp_in_level: number;
  xp_to_next_level: number;
  streak_days: number;
  streak_active_today: boolean;
  today_work_minutes: number;
  today_work_minutes_goal: number;
  is_today_perfect: boolean;
}

export async function getUserStats(days = 7): Promise<UserStatsRead> {
  // tz_offset = minutes east of UTC, so the server can bucket events by the
  // caller's local day instead of UTC.
  const tz = new Date().getTimezoneOffset() * -1;
  return apiFetch<UserStatsRead>(`/stats?days=${days}&tz_offset=${tz}`);
}

export async function getUserXp(): Promise<UserProgressRead> {
  const tz = new Date().getTimezoneOffset() * -1;
  return apiFetch<UserProgressRead>(`/xp?tz_offset=${tz}`);
}

export interface LabelRelabelResult {
  updated_sessions: number;
  from_label: string;
  to_label: string;
}

/**
 * Rename or merge a focus (time-usage) label across all sessions.
 * Passing an existing label as `to` merges the two buckets; a new name
 * just renames. Applied globally on the server.
 */
export async function relabelFocusLabel(
  from: string,
  to: string,
): Promise<LabelRelabelResult> {
  return apiFetch<LabelRelabelResult>("/stats/labels/relabel", {
    method: "POST",
    body: JSON.stringify({ from_label: from, to_label: to }),
  });
}

File: backend/app/schemas/stats.py
"""Pydantic schemas for stats aggregates."""

from datetime import date

from pydantic import BaseModel, Field


class DailyTaskStat(BaseModel):
    date: date
    total: int
    done: int


class DailyMoodStat(BaseModel):
    date: date
    mood: str


class DailyWorkStat(BaseModel):
    """Minutes of work (pomodoro_work + stopwatch) on a given local day."""

    date: date
    minutes: int


class WorkLabelStat(BaseModel):
    """Minutes grouped by the focus attached to work sessions."""

    label: str
    minutes: int


class LabelRelabel(BaseModel):
    """Request to rename or merge a focus label globally.

    `to_label` equal to an existing label merges the two (their time
    records combine); a fresh name simply renames. The operation rewrites
    `work_label` on every matching past session so the change is permanent
    and shows up everywhere the label is reported.
    """

    from_label: str = Field(..., min_length=1, max_length=300)
    to_label: str = Field(..., min_length=1, max_length=300)


class LabelRelabelResult(BaseModel):
    updated_sessions: int
    from_label: str
    to_label: str


class ProjectTimeStat(BaseModel):
    """Minutes grouped by the project bucket the session's linked task
    belongs to. Sessions inherit project transitively via linked_task_id.
    `project_id` is None for the "(no project)" bucket."""

    project_id: int | None
    name: str
    minutes: int


class WeeklySummaryCounts(BaseModel):
    work_minutes: int = 0
    tasks_done: int = 0
    notes: int = 0
    feynman: int = 0
    moods: int = 0


class WeeklySummary(BaseModel):
    """Last 7 days vs the prior 7 days, in the caller's local tz."""

    this_week: WeeklySummaryCounts
    prev_week: WeeklySummaryCounts


class UserStatsRead(BaseModel):
    days: int
    daily_tasks: list[DailyTaskStat]
    daily_moods: list[DailyMoodStat]
    daily_work_minutes: list[DailyWorkStat]
    work_labels: list[WorkLabelStat]
    time_per_project: list[ProjectTimeStat] = []
    total_tasks_done: int
    total_work_minutes: int
    total_notes: int
    total_feynman: int
    total_moods: int = 0
    weekly_summary: WeeklySummary | None = None
